### PR TITLE
[JAX] Sharding Utils

### PR DESCRIPTION
**Pull request by mingxu1067**
# Description

Adding sharding utils, like getting rank or group of a specific device, for further development of parallelism.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

1. Adding a device list with all detected devices and a proper sharding.
2. Adding two functions to get rank and group ID of the given mesh axis and device ID.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by denera — DISMISSED** [review]
LGTM, pending CI.

File: transformer_engine/jax/sharding.py [hunk 1]
```diff
@@ -132,13 +132,16 @@ def get_padded_spec(spec, ndim):
     return spec + (None,) * (ndim - len(spec))
 
 
-def lax_paral_op(x: jnp.array, ops: Callable, mesh_resource: str):
+def lax_paral_op(x: jnp.array, ops: Callable, mesh_resource: str, kwargs: dict = None):
```
Comment (zlsh80826) [inline transformer_engine/jax/sharding.py:135]: Doesn't it `**kwargs`?

**Comment by zlsh80826** [comment]
Hi @mingxu1067, do you have documentation on how `group` and `size` are defined here? It seems they follow some rules, but I'm not understanding them well.

File: transformer_engine/jax/sharding.py (hunk 1 above)
Comment (mingxu1067) [reply transformer_engine/jax/sharding.py]: Done

**Comment by mingxu1067** [comment]
> Hi @mingxu1067, do you have documentation on how `group` and `size` are defined here? It seems they follow some rules, but I'm not understanding them well.

I followed the MPI concept to design the functionalities of `group` and `rank`. In short, `rank` is the local ID of devices in the given axis, that is to specify which the part of the device belongs to in the axis. For example, if we have `Mesh={'TP': 8}`, then `rank` of `TP` of GPU-0 is `0`. Similarly, GPU-1 is `1` and -2 is `2`. `group` is to specify which GPUs are in the same group to perform communications. For example, if we have `Mesh={'DP':2, TP:'4'}`, then GPU 0-3 is in the same TP group to perform `AllReduce`. Similarly,  GPU 4-7 is in the same TP group to perform `AllReduce`.

**Comment by mingxu1067** [comment]
As @zlsh80826's suggestion, `jax.lax.axis_index()` might be a in-built solution to obtain `rank`. Testing.
